### ISSUE TITLE
Suggest Usage of AsRef, Into, …

### ISSUE THREAD

**Issue by killercup** [Manishearth/rust-clippy#1114]
_And now for something completely crazy\* that just came to my mind\**…_

Given a function `f` that has an input parameter `x` of type `X`, it is possible to determine all method calls in the function body (and, recursively, all function that are called in the body with `x` as parameter).

If `X` is a concrete type (i.e., not generic), and all the methods are on traits that `X` implements, it may be possible to find some constraints `C` to replace `X` with a generic type `T: C` so that `f<T>(x)` returns the same result as `f<X>(x)` for all possible values of `x`, _but_ allows strictly more inputs.

**tl;dr** Instead of `fn foo(s: &str)` you could probably also write `fn foo<T: AsRef<str>>(s: T)` and some other, more complicated stuff.

- - -

Yes I know this will be slow and crazy and—look, I needed to go all science-y to even describe it, I know this is a long shot 😄 

- - -

\* So basically the same as all the other issues I opened here 😉 
\** I got the idea from https://github.com/cksac/fake-rs/pull/2 (which is probably even more complicated).

**Comment by mcarton** [comment]
We don't have a E-crazy tag so I'll fo for E-hard :smile:

**Comment by killercup** [comment]
> We don't have a E-crazy tag

I was hoping for _A-u-crazy_ 😄

**Comment by kamalmarhubi** [comment]
I'm interested in this lint, at least for the string cases. I'd also be interested in something concise to read on idiomatic usage here to get a better grasp.

**Comment by kamalmarhubi** [comment]
Also, the non-crazy version that just looks at `String` and `&str` and related conversions would go along way.

**Comment by oli-obk** [comment]
I think the go-to text is https://doc.rust-lang.org/book/borrow-and-asref.html

also the docs on `std::convert`: https://doc.rust-lang.org/std/convert/index.html

**Comment by kamalmarhubi** [comment]
@oli-obk so I'd been told in the past the `AsRef<str>` was less idiomatic than `&str` despite it being more ergonomic for the caller. Just started a urlo discussion that will hopefully help clarify: https://users.rust-lang.org/t/idiomatic-string-parmeter-types-str-vs-asref-str-vs-into-string/7934

**Comment by llogiq** [comment]
The problem with `AsRef<str>` vs. `&str` is the latter allows autoderef to pick up the type easily, whereas due to `impl AsRef<T> for T` using Strings with `AsRef` may either break inference or move them by accident.
